cpp/tests/groupby/host_udf_tests.cu

```diff
@@ -0,0 +1,547 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf_test/base_fixture.hpp>
+#include <cudf_test/column_wrapper.hpp>
+#include <cudf_test/debug_utilities.hpp>
+#include <cudf_test/iterator_utilities.hpp>
+#include <cudf_test/type_lists.hpp>
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/aggregation/aggregation.hpp>
+#include <cudf/detail/null_mask.hpp>
+#include <cudf/detail/valid_if.cuh>
+#include <cudf/groupby.hpp>
+#include <cudf/reduction.hpp>
+#include <cudf/scalar/scalar_factories.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+#include <rmm/device_uvector.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/iterator/zip_iterator.h>
+#include <thrust/transform.h>
+#include <thrust/transform_reduce.h>
+
+/**
+ * @brief A host-based UDF implementation.
+ *
+ * The aggregations perform the following computation:
+ *  - For reduction: compute `sum(value^2, for value in group)` (this is sum of squared).
+ *  - For segmented reduction: compute `segment_size * sum(value^2, for value in group)`.
+ *  - For groupby: compute `(group_idx + 1) * sum(value^2, for value in group)`.
+ *
+ * In addition, for segmented reduction, if null_policy is set to `INCLUDE`, the null values are
+ * replaced with an initial value if it is provided.
+ */
+template <typename cudf_aggregation>
+struct test_udf_simple_type : cudf::host_udf_base {
+  static_assert(std::is_same_v<cudf_aggregation, cudf::reduce_aggregation> ||
+                std::is_same_v<cudf_aggregation, cudf::segmented_reduce_aggregation> ||
+                std::is_same_v<cudf_aggregation, cudf::groupby_aggregation>);
+
+  test_udf_simple_type() = default;
+
+  [[nodiscard]] std::unordered_set<input_kind> const& get_required_data_kinds() const override
```

Comment:
This design requires users to ensure that their custom hash set is static. Instead, could we ask users to pass it to the constructor and then assign it to a data member?

Reply:
Yes that is also one of the possible approaches. However, I would prefer to have an accessor this way (will modify  the interface to remove `const&`) so the libcudf class `cudf::detail::host_udf_aggregation` only stores one member variable (the unique pointer to base class host_udf). That approach seems to be better in term of encapsulation.